File: binaryTree.hs
import Data.Time (getCurrentTime, diffUTCTime, UTCTime)
import System.Mem (performGC)
import GHC.Stats (getRTSStatsEnabled, getRTSStats, GCDetails(..), RTSStats(..))
import Text.Printf (printf)
import Control.DeepSeq (NFData (rnf), deepseq)
import System.Random (randomRIO)

-- Binary Tree
data BinaryTree a = Empty
                  | Node a (BinaryTree a) (BinaryTree a)
                  deriving (Show, Eq)

instance NFData a => NFData (BinaryTree a) where
  rnf Empty = ()
  rnf (Node x left right) = rnf x `seq` rnf left `seq` rnf right


-- Insert
insert :: Ord a => a -> BinaryTree a -> BinaryTree a
insert x Empty = Node x Empty Empty
insert x (Node y left right)
    | x < y     = Node y (insert x left) right
    | otherwise = Node y left (insert x right)

-- Search
search :: Ord a => a -> BinaryTree a -> Bool
search _ Empty = False
search x (Node y left right)
    | x == y    = True
    | x < y     = search x left
    | otherwise = search x right

-- Find the minimum value
findMin :: BinaryTree a -> a
findMin (Node x Empty _) = x
findMin (Node _ left _)  = findMin left
findMin Empty = error "Empty tree has no minimum"

-- Remove
remove :: Ord a => a -> BinaryTree a -> BinaryTree a
remove _ Empty = Empty
remove x (Node y left right)
    | x < y     = Node y (remove x left) right
    | x > y     = Node y left (remove x right)
    | otherwise = case (left, right) of
                    (Empty, Empty) -> Empty
                    (Empty, _)     -> right
                    (_, Empty)     -> left
                    _              -> let minValue = findMin right
                                       in Node minValue left (remove minValue right)

-- Traversals (in-order, pre-order, post-order)
inOrder :: BinaryTree a -> [a]
inOrder Empty = []
inOrder (Node x left right) = inOrder left ++ [x] ++ inOrder right

preOrder :: BinaryTree a -> [a]
preOrder Empty = []
preOrder (Node x left right) = [x] ++ preOrder left ++ preOrder right

postOrder :: BinaryTree a -> [a]
postOrder Empty = []
postOrder (Node x left right) = postOrder left ++ postOrder right ++ [x]

-- timing and memory
measureTime :: IO a -> IO Double
measureTime action = do
    start <- getCurrentTime
    result <- action
    end <- getCurrentTime
    return $ realToFrac (diffUTCTime end start) * 1e9 -- Convert to nanoseconds

measureMemory :: IO a -> IO Integer
measureMemory action = do
    performGC
    statsBefore <- getRTSStats
    _ <- action
    performGC
    statsAfter <- getRTSStats
    let liveBytesBefore = fromIntegral $ gcdetails_live_bytes (gc statsBefore)
        liveBytesAfter = fromIntegral $ gcdetails_live_bytes (gc statsAfter)
    return $ liveBytesAfter - liveBytesBefore


printPretty :: String -> Double -> String -> IO ()
printPretty task duration unit = printf "%s: %.0f %s\n" task duration unit

-- function to generate lists of n random elements
generateRandomNumbers :: Int -> IO [Int]
generateRandomNumbers n = mapM (\_ -> randomRIO (1, 1000000)) [1..n] 

-- Main
main :: IO ()
main = do
    -- Define n !!!!!!!
    let n = 10000000
    elements <- generateRandomNumbers n  
    putStrLn $ "\nNumber of elements: " ++ show n

    -- Insert 
    insertTime <- measureTime $ do
        let tree = foldr insert Empty elements
        tree `deepseq` return tree 
    printPretty "\nTime taken to insert elements" insertTime "nanoseconds"

    -- Create the tree
    let tree = foldr insert Empty elements

    -- Measure memory used by the tree
    treeMemory <- measureMemory $ tree `deepseq` return tree
    printPretty "Memory used by the tree" (fromIntegral treeMemory) "bytes"

    -- In-order traversal
    inOrderTime <- measureTime $ do
        let result = inOrder tree
        result `deepseq` return result
    inOrderMemory <- measureMemory $ let result = inOrder tree in result `deepseq` return result
    printPretty "\nTime taken for in-order traversal" inOrderTime "nanoseconds"
    printPretty "Auxiliary memory for in-order traversal" (fromIntegral inOrderMemory) "bytes"

    -- Pre-order traversal
    preOrderTime <- measureTime $ do
        let result = preOrder tree
        result `deepseq` return result
    preOrderMemory <- measureMemory $ let result = preOrder tree in result `deepseq` return result
    printPretty "\nTime taken for pre-order traversal" preOrderTime "nanoseconds"
    printPretty "Auxiliary memory for pre-order traversal" (fromIntegral preOrderMemory) "bytes"

    -- Post-order traversal
    postOrderTime <- measureTime $ do
        let result = postOrder tree
        result `deepseq` return result
    postOrderMemory <- measureMemory $ let result = postOrder tree in result `deepseq` return result
    printPretty "\nTime taken for post-order traversal" postOrderTime "nanoseconds"
    printPretty "Auxiliary memory for post-order traversal" (fromIntegral postOrderMemory) "bytes"

    -- Search for an element
    searchTime <- measureTime $ do
        let result = search (n + 1) tree
        result `deepseq` return result
    printPretty "\nTime taken to search for an element" searchTime "nanoseconds"

    -- Remove an element
    removeTime <- measureTime $ do
        let treeAfterRemoval = remove (n `div` 2) tree
        treeAfterRemoval `deepseq` return treeAfterRemoval
    printPretty "\nTime taken to remove an element" removeTime "nanoseconds"

File: binaryTree.py
import random
import time  # for time complexity
import sys  # for space complexity
import tracemalloc  # for auxiliary memory

# Calculate the memory of a node
def node_memory(node):
    if not node:
        return 0
    return sys.getsizeof(node) + node_memory(node.left) + node_memory(node.right)

# Node class
class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

# Binary Tree class
class BinaryTree:

    # Constructor
    def __init__(self):
        self.root = None

    # Insert method
    def insert(self, key):
        if not self.root:
            self.root = Node(key)
        else:
            self._insert(self.root, key)

    def _insert(self, current, key):
        if key < current.key:
            if current.left:
                self._insert(current.left, key)
            else:
                current.left = Node(key)
        else:
            if current.right:
                self._insert(current.right, key)
            else:
                current.right = Node(key)

    # Search method
    def search(self, key):
        return self._search(self.root, key)

    def _search(self, current, key):
        if not current:
            return False
        if current.key == key:
            return True
        elif key < current.key:
            return self._search(current.left, key)
        else:
            return self._search(current.right, key)

    # Remove method
    def remove(self, key):
        self.root = self._remove(self.root, key)

    def _remove(self, current, key):
        if not current:
            return current
        if key < current.key:
            current.left = self._remove(current.left, key)
        elif key > current.key:
            current.right = self._remove(current.right, key)
        else:
            if not current.left:
                return current.right
            elif not current.right:
                return current.left
            min_larger_node = self._find_min(current.right)
            current.key = min_larger_node.key
            current.right = self._remove(current.right, min_larger_node.key)
        return current

    def _find_min(self, current):
        while current.left:
            current = current.left
        return current

    # In-order traversal
    def inorder_traversal(self):
        result = []
        self._inorder(self.root, result)
        return result

    def _inorder(self, current, result):
        if current:
            self._inorder(current.left, result)
            result.append(current.key)
            self._inorder(current.right, result)

    # Pre-order traversal
    def preorder_traversal(self):
        result = []
        self._preorder(self.root, result)
        return result

    def _preorder(self, current, result):
        if current:
            result.append(current.key)
            self._preorder(current.left, result)
            self._preorder(current.right, result)

    # Post-order traversal
    def postorder_traversal(self):
        result = []
        self._postorder(self.root, result)
        return result

    def _postorder(self, current, result):
        if current:
            self._postorder(current.left, result)
            self._postorder(current.right, result)
            result.append(current.key)

# Main execution
if __name__ == "__main__":
    bt = BinaryTree()
    n = 10000  # Number of elements
    print(f"\nNumber of elements: {n}")

    # Time for inserts
    start = time.time()
    for i in range(n - 1):
        bt.insert(random.randint(1, 999999))
    # Insert the n-th element to be removed later
    bt.insert(872)
    end = time.time()
    print(f"\nTime taken to insert {n} elements: {(end - start):.5f} seconds")

    # Measure tree memory
    tree_memory = node_memory(bt.root)
    print(f"Memory used by the tree: {tree_memory} bytes")

    # Time and auxiliary memory for in-order traversal
    start = time.time()
    tracemalloc.start()
    bt.inorder_traversal()
    end = time.time()
    current, peak = tracemalloc.get_traced_memory()
    print(f"\nAuxiliary memory for in-order traversal: {current} bytes")
    tracemalloc.stop()
    print(f"Time taken for in-order traversal of {n} elements: {(end - start):.5f} seconds")

    # Time and auxiliary memory for pre-order traversal
    start = time.time()
    tracemalloc.start()
    bt.preorder_traversal()
    end = time.time()
    current, peak = tracemalloc.get_traced_memory()
    print(f"\nAuxiliary memory for in-order traversal: {current} bytes")
    tracemalloc.stop()
    print(f"Time taken for pre-order traversal of {n} elements: {(end - start):.5f} seconds")

    # Time and auxiliary memory for post-order traversal
    start = time.time()
    tracemalloc.start()
    bt.postorder_traversal()
    end = time.time()
    current, peak = tracemalloc.get_traced_memory()
    print(f"\nAuxiliary memory for in-order traversal: {current} bytes")
    tracemalloc.stop()
    print(f"Time taken for post-order traversal of {n} elements: {(end - start):.5f} seconds")

    # Time to search an element (worst case scenario)
    start = time.time()
    print(f"\nSearch 88901: {bt.search(88901)}")
    end = time.time()
    print(f"Time taken to search for 88901: {(end - start):.5f} seconds")

    # Time to remove an element (worst case scenario)
    start = time.time()
    bt.remove(872)
    end = time.time()
    print(f"\nTime taken to remove 872: {(end - start):.5f} seconds")
